dqc_us_rules/dqc_us_0041.py

```diff
@@ -0,0 +1,294 @@
+        into memory
+    :rtype: None
+    """
+    val.linroleDefinitionIsDisclosure = (
+        re.compile(r"-\s+Disclosure\s+-\s", re.IGNORECASE)
+    )
+
+    val.linkroleDefinitionStatementSheet = (
+        re.compile(r"[^-]+-\s+Statement\s+-\s+.*", re.IGNORECASE)
+    )  # no restriction to type of statement
+
+    val.ugtNamespace = None
+    cntlr = val.modelXbrl.modelManager.cntlr
+
+    year = 2014
+    for ugt in ugtDocs:
+        ugt_default_dimensions_json_file = os.path.join(
+            os.path.dirname(__file__),
+            'resources',
+            'DQC_US_0041',
+            "_".join([str(year), "ugt-default-dimensions.json"])
+        )
+
+        _make_cache(
+            val,
+            ugt,
+            cntlr,
+            ugt_default_dimensions_json_file
+        )
+
+        year += 1
+
+
+def _is_in_namespace(val, ugt_namespace):
+    """
+    Returns true if the ugt_namespace is the same as the current vals namespace
+
+    :param val: ValidateXbrl that is being validated
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl
+    :param ugt_namespace: Namespace of taxonomy for current year
+    :type ugt_namespace: str
+    :return: True if ugt_namespace is in the val.modelXbrl's namespace docs
+    :rtype: bool
+    """
+    return (ugt_namespace in val.modelXbrl.namespaceDocs and
+            len(val.modelXbrl.namespaceDocs[ugt_namespace]) > 0
+            )
+
+
+def _load_cache(val):
+    """
+    Loads the cached taxonomy default demensions. If the file isn't cached yet
+    it will create a new cache
+
+    :param val: ValidateXbrl to be validated
+    :type val: :class: '~arelle.ValidateXbrl.ValidateXbrl'
+    :return: no explicit return, but loads caches for dqc_us_0041
+    :rtype: None
+    """
+    val.linroleDefinitionIsDisclosure = (
+        re.compile(r"-\s+Disclosure\s+-\s", re.IGNORECASE)
+    )
+
+    val.linkroleDefinitionStatementSheet = (
+        re.compile(r"[^-]+-\s+Statement\s+-\s+.*", re.IGNORECASE)
+    )  # no restriction to type of statement
+
+    val.ugtNamespace = None
+    cntlr = val.modelXbrl.modelManager.cntlr
+
+    year = 2014
+
+    for ugt in ugtDocs:
+        ugt_namespace = ugt["namespace"]
+        if _is_in_namespace(val, ugt_namespace):
+            ugt_default_dimensions_json_file = os.path.join(
+                os.path.dirname(__file__),
+                'resources',
+                'DQC_US_0041',
+                "_".join([str(year), "ugt-default-dimensions.json"])
+            )
+
+            file = None
+
+            try:
+                file = openFileStream(
+                       cntlr,
+                       ugt_default_dimensions_json_file,
+                       'rt',
+                       encoding='utf-8'
+                )
+
+                val.usgaapDefaultDimensions = json.load(file)
+                file.close()
+
+            except FileNotFoundError:
+                if file:
+                    file.close()
+        year += 1
```

Comment:
Extracted start year out to constant, but since I am already iterating over list so I couldn't take your suggestion